Utils/ExtensionUtils.swift

```diff
@@ -8,33 +8,41 @@ import MobileCoreServices
+    /// future code can possibly interact with a web page to find a proper icon.
     public static func extractSharedItemFromExtensionContext(extensionContext: NSExtensionContext?, completionHandler: (ShareItem?, NSError!) -> Void) {
-        if extensionContext != nil {
-            if let inputItems : [NSExtensionItem] = extensionContext!.inputItems as? [NSExtensionItem] {
-                for inputItem in inputItems {
-                    if let attachments = inputItem.attachments as? [NSItemProvider] {
-                        for attachment in attachments {
-                            if attachment.hasItemConformingToTypeIdentifier(kUTTypeURL as String) {
-                                attachment.loadItemForTypeIdentifier(kUTTypeURL as String, options: nil, completionHandler: { (obj, err) -> Void in
-                                    if err != nil {
-                                        completionHandler(nil, err)
-                                    } else {
-                                        let title = inputItem.attributedContentText?.string as String?
-                                        if let url = obj as? NSURL {
-                                            completionHandler(ShareItem(url: url.absoluteString, title: title, favicon: nil), nil)
-                                        } else {
-                                            completionHandler(nil, NSError(domain: "org.mozilla.fennec", code: 999, userInfo: ["Problem": "Non-URL result."]))
-                                        }
-                                    }
-                                })
-                                return
-                            }
+
+        guard let extensionContext = extensionContext,
+              let inputItems = extensionContext.inputItems as? [NSExtensionItem] else {
+            completionHandler(nil, nil)
+            return
+        }
+
+        for inputItem in inputItems {
+            guard let attachments = inputItem.attachments as? [NSItemProvider] else { continue }
+
+            for attachment in attachments {
+                if attachment.hasItemConformingToTypeIdentifier(kUTTypeURL as String) {
+                    attachment.loadItemForTypeIdentifier(kUTTypeURL as String, options: nil, completionHandler: { obj, err in
+                        guard err == nil else {
+                            completionHandler(nil, err)
+                            return
                         }
-                    }
+
+                        guard let url = obj as? NSURL else {
+                            completionHandler(nil, NSError(domain: "org.mozilla.fennec", code: 999, userInfo: ["Problem": "Non-URL result."]))
+                            return
+                        }
+
+                        let title = inputItem.attributedContentText?.string
+                        completionHandler(ShareItem(url: url.absoluteString, title: title, favicon: nil), nil)
```

Comment:
The `completionHandler` is called for every element in `inputItems`. Is this right? Doesn't sound like a completion handler if it is right.